src/main/java/org/hisp/dhis/antlr/AntlrParserUtils.java

```diff
@@ -188,106 +188,132 @@ public static Date castDate( Object object )
      */
     public static Object castClass( Class<?> clazz, Object object )
     {
-        if ( object instanceof Double )
+        if ( object == null )
         {
-            return castDoubleObject( (Double) object, clazz );
+            return null;
         }
 
-        if ( object instanceof String )
+        if ( clazz == Double.class && makeDouble( object ) != null )
         {
-            return castStringObject( (String) object, clazz );
+            return makeDouble( object );
         }
 
-        if ( object instanceof Boolean )
+        if ( clazz == Boolean.class && makeBoolean( object ) != null )
         {
-            return castBooleanObject( (Boolean) object, clazz );
+            return makeBoolean( object );
         }
 
-        try
+        if ( clazz == Date.class && makeDate( object ) != null )
         {
-            return clazz.cast( object );
+            return makeDate( object );
         }
-        catch ( Exception e )
+
+        if ( clazz == String.class )
         {
-            throw new ParserExceptionWithoutContext( "Could not cast value to " + clazz.getSimpleName() );
+            return makeString( object );
         }
+
+        throw new ParserExceptionWithoutContext( "Could not cast " + object.getClass().getSimpleName() +
```

Comment:
Will this exception ever thrown?
Maybe it is safe to have this here anyway but we know that AntlrOperatorCompare will never arrive here.
So my other question is, does it make sense that no matter what we are back falling to String comparison?
I do not the answer, maybe it is valid to do so. I am just thinking if the user really make a mistake in the expression and he thinks he is comparing something and we just give him a result based on strings.

Reply:
Agreed, we know that `AntlrOperatorCompare` will never get here, but this is a public method that is also called by the DHIS2 backend when it requires that an expression produce a certain type. This is the last statement in a method that returns something, so we either need to return something or throw an exception. Throwing an exception seems like a better choice if we get a class we don't recognize. :)

.  .  .

Good point about converting everything. I agree that it might be more useful to throw a type error than to accept something that presumably won't work at runtime.

The intent of this PR is to make consistent between the left and right sides of a comparison the type conversions that we already had. We used to only look at the left side for type and then see if the right side could be cast to match. So we used to compare 10 > "2" as numbers (true), but "2" < 10 as strings (false). We used to compare true == 1 by converting 1 to a Boolean, but we threw an error on 1 == true because we couldn't convert true to a number. We used to allow "abc" > 1 because we could convert 1 to a string, but throw an error on 1 < "abc" because we couldn't convert "abc" to a number.

But it also went one step further as you point out, falling back to a string comparison for everything. I've now changed this (see commit [ac4c6a2)](https://github.com/dhis2/dhis2-antlr-expression-parser/pull/30/commits/ac4c6a249423441bc296a3ede7f30276667b1e11) so that it only does type conversions when one side or the other has the type that we could use for comparison. So we don't compare 2.5 == false by converting them both to strings. Also we don't compare the strings "2" < "10" as if they are both numeric. (I had to adjust some of the test cases accordingly.)

I still have some reservations about converting as much as we do. I was in total agreement with Zubair's initial comment in [DHIS2-12070](https://jira.dhis2.org/browse/DHIS2-12070). It looked like a user type error to me too, and I thought it was reasonable to return an error. But instead we "fixed" it by allowing numbers (well, at least integers) to have a boolean value depending on whether they were zero or not. I think I would have preferred not to do this conversion from integer to boolean, but some languages (like C) do it. I assumed that there was pressure to make it work like before, when apparently we didn't generate an error. I'm guessing that the more strict type-checking was a result of my code in [DHIS2-10818](https://jira.dhis2.org/browse/DHIS2-10818) to support database values other than numeric. @zubaira, any further comment on this?

Thanks for the careful review. Any further thoughts?

Reply:
Just to summarise, we are going to have an error only for comparison like this:
`3.1 > true`
Is that correct?
If this is the case maybe it is good to have a test just to highlight that

Reply:
Uh, yeah, I guess you're right. I'm regretting that we've been so generous with type conversions.

I've added the test. Thanks.